src/main/java/uk/gov/justice/digital/delius/controller/secure/CustodyKeyDatesController.java

```diff
@@ -75,19 +76,15 @@ public Custody replaceAllCustodyKeyDateByNomsNumberAndBookingNumber(final @PathV
                                                                         final @PathVariable String bookingNumber,
                                                                         final @RequestBody ReplaceCustodyKeyDates replaceCustodyKeyDates) {
         log.info("Call to replaceAllCustodyKeyDateByNomsNumberAndBookingNumber for {} booking {} with dates {}", nomsNumber, bookingNumber, replaceCustodyKeyDates);
-        return Custody
-                .builder()
-                .keyDates(CustodyRelatedKeyDates
-                        .builder()
-                        .conditionalReleaseDate(replaceCustodyKeyDates.getConditionalReleaseDate())
-                        .expectedReleaseDate(replaceCustodyKeyDates.getExpectedReleaseDate())
-                        .hdcEligibilityDate(replaceCustodyKeyDates.getHdcEligibilityDate())
-                        .licenceExpiryDate(replaceCustodyKeyDates.getLicenceExpiryDate())
-                        .paroleEligibilityDate(replaceCustodyKeyDates.getParoleEligibilityDate())
-                        .postSentenceSupervisionEndDate(replaceCustodyKeyDates.getPostSentenceSupervisionEndDate())
-                        .sentenceExpiryDate(replaceCustodyKeyDates.getSentenceExpiryDate())
-                        .build())
-                .build();
+
+        var offenderId = offenderService.offenderIdOfNomsNumber(nomsNumber).orElseThrow(() -> new NotFoundException(String.format("Offender with NOMS number %s not found", nomsNumber)));
+        try {
+            var convictionId = convictionService.getSingleActiveConvictionIdByOffenderIdAndPrisonBookingNumber(offenderId, bookingNumber).orElseThrow(() -> new NotFoundException(String.format("Conviction with bookingNumber %s not found for offender with NOMS number %s", bookingNumber, nomsNumber)));
+            return convictionService.addOrReplaceOrDeleteCustodyKeyDates(offenderId, convictionId, replaceCustodyKeyDates);
+        } catch (DuplicateConvictionsForBookingNumberException e) {
```

Comment:
No 100% happy with this pattern of handling exceptions here but it is consistent with the rest of this controller (and of course does work)